### PR TITLE
Fix #83045 by moving some crate loading verification code to a better place

### LINKED ISSUE #83045: ICE: compiler/rustc_metadata/src/creader.rs:136:32: Failed to get crate data for crate22
### Code

This error was found when building [clitrans](https://github.com/wfxr/clitrans) on macos by github CI ([logs](https://github.com/wfxr/clitrans/runs/2091662556?check_suite_focus=true)). Code may be irrelevant because Ubuntu and Windows target were both successfully built and the first commit(https://github.com/wfxr/clitrans/commit/57ca835c2e009168f29e21208c3b01a9cc1809bd) that caused this error only did some refactors on `Cargo.toml`, no code changes.

**EDIT:**

After I revert the `Cargo.toml` changes (https://github.com/wfxr/clitrans/commit/57ca835c2e009168f29e21208c3b01a9cc1809bd) the error is gone.

### Meta

`rustc --version --verbose`:
```
rustc 1.52.0-nightly (4a8b6f708 2021-03-11)
binary: rustc
commit-hash: 4a8b6f708c38342a6c74aa00cf4323774c7381a6
commit-date: 2021-03-11
host: x86_64-apple-darwin
release: 1.52.0-nightly
LLVM version: 12.0.0
```

### Error output

```
thread 'rustc' panicked at 'Failed to get crate data for crate22', compiler/rustc_metadata/src/creader.rs:136:32
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.52.0-nightly (4a8b6f708 2021-03-11) running on x86_64-apple-darwin

note: compiler flags: -C opt-level=3 -C linker-plugin-lto -C codegen-units=1 --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `clitrans`

To learn more, run the command again with --verbose.
Error: Process completed with exit code 101.
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0: _rust_begin_unwind
   1: std::panicking::begin_panic_fmt
   2: rustc_metadata::creader::CStore::get_crate_data::{{closure}}
   3: rustc_metadata::creader::CrateLoader::maybe_resolve_crate
   4: rustc_metadata::creader::CrateLoader::maybe_process_path_extern
   5: rustc_resolve::Resolver::extern_prelude_get
   6: rustc_resolve::macros::<impl rustc_resolve::Resolver>::early_resolve_ident_in_lexical_scope
   7: rustc_resolve::Resolver::resolve_path_with_ribs::{{closure}}
   8: rustc_resolve::Resolver::resolve_path_with_ribs
   9: rustc_resolve::imports::ImportResolver::resolve_imports
  10: rustc_resolve::macros::<impl rustc_expand::base::ResolverExpand for rustc_resolve::Resolver>::resolve_imports
  11: rustc_expand::expand::MacroExpander::fully_expand_fragment
  12: rustc_expand::expand::MacroExpander::expand_crate
  13: rustc_session::utils::<impl rustc_session::session::Session>::time
  14: rustc_interface::passes::configure_and_expand_inner
  15: rustc_interface::passes::configure_and_expand::{{closure}}
  16: rustc_data_structures::box_region::PinnedGenerator<I,A,R>::new
  17: rustc_interface::passes::configure_and_expand
  18: rustc_interface::queries::Queries::expansion
  19: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  20: rustc_span::with_source_map
  21: rustc_interface::interface::create_compiler_and_run
  22: rustc_span::with_session_globals
```

</p>
</details>

### PR DESCRIPTION
**Pull request by michaelwoerister**
r? @eddyb

### CODE REVIEW

**Review by eddyb — APPROVED** [review]

File: src/test/run-make/issue-83045/Makefile [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+all:
+	$(RUSTC) --crate-name=a --crate-type=rlib a.rs --verbose
+	$(RUSTC) --crate-name=b --crate-type=rlib --extern a=$(TMPDIR)/liba.rlib b.rs --verbose
+	# NOTE: We use $(BARE_RUSTC) below so that the compiler can't find liba.rlib
+	#       If we used $(RUSTC) the additional `-L $(TMPDIR)` option would allow rustc to
+	#       actually find the crate.
+	$(BARE_RUSTC) --out-dir $(TMPDIR) --extern b=$(TMPDIR)/libb.rlib --crate-type=rlib \
+	    --edition=2018 c.rs  &> $(TMPDIR)/output.txt || exit 0
+	# Check that we get the expected error message
+	$(CGREP) "error[E0463]" < $(TMPDIR)/output.txt
```
Comment (eddyb) [inline src/test/run-make/issue-83045/Makefile:26]: You probably need backslashes for the `[` and `]`?
Might be easier to match on the message itself which is:
```
can't find crate for `a` which `b` depends on
```

**Comment by eddyb** [comment]
r=me with CI fixed

File: src/test/run-make/issue-83045/Makefile (hunk 1 above)
Comment (michaelwoerister) [reply src/test/run-make/issue-83045/Makefile:26]: I'll just match on `E0463`. That should be sufficient.

**Comment by michaelwoerister** [comment]
I don't understand what's going on here. The string to be matched is right there in the output and the test passes for me locally `:(`

**Comment by michaelwoerister** [comment]
The comment from inside the Makefile somehow ends up in `output.txt`?

**Comment by michaelwoerister** [comment]
OK, piping the compiler's output through `tee` seems to have fixed the test case. I don't even want to know...

@bors r=eddyb

**Pull request merged**